File: app/src/main/java/za/org/grassroot2/presenter/AroundMePresenter.java
package za.org.grassroot2.presenter;

import android.location.Location;
import android.support.annotation.NonNull;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import za.org.grassroot2.database.DatabaseService;
import za.org.grassroot2.model.AroundItem;
import za.org.grassroot2.model.Group;
import za.org.grassroot2.presenter.fragment.BaseFragmentPresenter;
import za.org.grassroot2.services.LocationManager;
import za.org.grassroot2.services.NetworkService;
import za.org.grassroot2.view.FragmentView;


public class AroundMePresenter extends BaseFragmentPresenter<AroundMePresenter.AroundMeView> {

    private LocationManager locationManager;
    private DatabaseService dbService;
    private NetworkService  networkService;
    private double      testLat           = -26.1925350;
    private double      testLong          = 28.0373235;
    static final double OFFSET = 0.00002;
    private Set<String> existingLocations = new HashSet<>();

    @Inject
    public AroundMePresenter(LocationManager manager, NetworkService networkService, DatabaseService dbService) {
        locationManager = manager;
        this.networkService = networkService;
        this.dbService = dbService;
    }

    @Override
    public void detach(AroundMeView view) {
        super.detach(view);
        locationManager.disconnect();
    }

    public void getCurrentLocation() {
        Location l = new Location("");
        l.setLatitude(testLat);
        l.setLongitude(testLong);
        locationManager.getCurrentLocation().subscribe(location -> view.renderLocation(l), Throwable::printStackTrace);
    }

    @Override
    public void onViewCreated() {
    }

    public void loadItemsAround() {
        networkService.getAllAround(testLat, testLong, 10000).flatMap(aroundItems -> {
            adjustLocations(aroundItems);
            return Observable.just(aroundItems);
        }).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe(itemsAround -> view.renderItemsAround(itemsAround), Throwable::printStackTrace);
    }

    private void adjustLocations(List<AroundItem> aroundItems) {
        int i = 0;
        for (AroundItem item : aroundItems) {
            String location = getLocationAsString(item);
            if (existingLocations.contains(location)) {
                item.setLatitude(item.getLatitude() + i * OFFSET);
                item.setLongitude(item.getLongitude() + i * OFFSET);
                location = getLocationAsString(item);
                i++;
            }
            existingLocations.add(location);
        }
    }

    @NonNull
    private String getLocationAsString(AroundItem item) {
        return String.valueOf(item.getLatitude()) + String.valueOf(item.getLongitude());
    }

    public void openGroupDetail(String uid) {
        dbService.load(Group.class, uid).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe(group -> {
            view.openGroupDetails(group.getUid());
        }, Throwable::printStackTrace);
    }

    public interface AroundMeView extends FragmentView {
        void renderLocation(Location location);
        void renderItemsAround(List<AroundItem> liveWireAlerts);
        void openGroupDetails(String uid);
    }

}

File: app/src/main/java/za/org/grassroot2/presenter/AroundMePresenter.kt
package za.org.grassroot2.presenter

import android.location.Location

import java.util.HashSet

import javax.inject.Inject

import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import za.org.grassroot2.database.DatabaseService
import za.org.grassroot2.model.AroundEntity
import za.org.grassroot2.model.Group
import za.org.grassroot2.presenter.fragment.BaseFragmentPresenter
import za.org.grassroot2.services.LocationManager
import za.org.grassroot2.services.NetworkService
import za.org.grassroot2.view.FragmentView


class AroundMePresenter @Inject
constructor(private val locationManager: LocationManager, private val networkService: NetworkService, private val dbService: DatabaseService) : BaseFragmentPresenter<AroundMePresenter.AroundMeView>() {
    private val testLat = -26.1925350
    private val testLong = 28.0373235
    private val existingLocations = HashSet<String>()

    override fun detach(view: AroundMeView) {
        super.detach(view)
        locationManager.disconnect()
    }

    fun getCurrentLocation() {
        val l = Location("")
        l.latitude = testLat
        l.longitude = testLong
        locationManager.currentLocation.subscribe({ location -> view.renderLocation(l) }, {t -> t.printStackTrace() })
    }

    override fun onViewCreated() {}

    fun loadItemsAround() {
        networkService.getAllAround(testLat, testLong, 10000).flatMap { aroundItems ->
            adjustLocations(aroundItems)
            Observable.just(aroundItems)
        }.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe({ itemsAround -> view.renderItemsAround(itemsAround) }, {t -> t.printStackTrace() })
    }

    private fun adjustLocations(aroundItems: List<AroundEntity>) {
        aroundItems.forEachIndexed { index, item ->
            var location = getLocationAsString(item)
            if (existingLocations.contains(location)) {
                item.latitude = item.latitude!! + index * OFFSET
                item.longitude = item.longitude!! + index * OFFSET
                location = getLocationAsString(item)
            }
            existingLocations.add(location)
        }
    }

    private fun getLocationAsString(item: AroundEntity): String {
        return item.latitude.toString() + item.longitude.toString()
    }

    fun openGroupDetail(uid: String) {
        dbService.load(Group::class.java, uid).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe({ group -> view.openGroupDetails(group.uid) }, {t -> t.printStackTrace() })
    }

    interface AroundMeView : FragmentView {
        fun renderLocation(location: Location)
        fun renderItemsAround(aroundItems: List<AroundEntity>)
        fun openGroupDetails(uid: String)
    }

    companion object {
        internal val OFFSET = 0.00002
    }

}
